### PR TITLE
Map basic testing

### PR DESCRIPTION
**Pull request by daviddrysdale**
In progress, but may be helpful given hashing discussions.

### CODE REVIEW

File: integration/map.go [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+	key0 = []byte{
+		0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10,
+		0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f, 0x20,
+	}
+	key1 = []byte{
+		0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f, 0x20,
+		0x21, 0x22, 0x23, 0x24, 0x25, 0x26, 0x27, 0x28, 0x29, 0x2a, 0x2b, 0x2c, 0x2d, 0x2e, 0x2f, 0x30,
+	}
+	value0 = []byte("value-0000")
+	value1 = []byte("value-0001")
```
Comment (golangcibot) [inline integration/map.go]: `value1` is unused

File: integration/map.go [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+	value2 = []byte("value-0002")
```
Comment (golangcibot) [inline integration/map.go]: `value2` is unused

File: integration/maptest/map.go [hunk 3]
```diff
@@ -105,16 +105,27 @@ func verifyGetMapLeavesResponse(mapVerifier *client.MapVerifier, getResp *trilli
...
 		leafHash := incl.GetLeaf().GetLeafHash()
 
-		wantLeafHash, err := mapVerifier.Hasher.HashLeaf(mapVerifier.MapID, index, leaf)
+		wantLeafHash, err := mapVerifier.Hasher.HashLeaf(mapVerifier.MapID, index, value)
 		if err != nil {
 			return err
 		}
-		if got, want := leafHash, wantLeafHash; !bytes.Equal(got, want) {
-			return fmt.Errorf("HashLeaf(%s): %x, want %x", leaf, got, want)
+		if len(value) == 0 {
```
Comment (golangcibot) [inline integration/maptest/map.go]: empty branch

File: merkle/map_verifier.go [hunk 4]
```diff
@@ -42,16 +43,28 @@ func VerifyMapInclusionProof(treeID int64, index, leaf, expectedRoot []byte, pro
...
+			return fmt.Errorf("leaf.LeafHash=%x matches neither HashLeaf()=%x nor HashEmpty()=%x", leaf.LeafHash, leafHash, emptyHash)
 		}
-		runningHash = leafHash
+		// The leaf value corresponds to a never-existed key; the calculations below use
+		// an empty hash to mark this case.
+		leafHash = nil
 	}
 
-	nID := storage.NewNodeIDFromHash(index)
+	var runningHash []byte
```
Comment (golangcibot) [inline merkle/map_verifier.go]: should merge variable declaration with assignment on next line

**Comment by daviddrysdale** [comment]
￼The commits "mapserver: deal with non-existent leaves", "map: cope with empty/cleared leaf values", and "merkle: clarify LeafHash() use" have my current understanding of how hashing should work...

File: gometalinter.json [hunk 5]
```diff
@@ -20,6 +20,6 @@
     ".+\\.pb\\.gw\\.go",
     "mock_.+\\.go"
   ],
-  "Cyclo": 25,
+  "Cyclo": 40,
```
Comment (gdbelvin) [inline gometalinter.json]: The cyclo check is designed to make sure we don't have overly complex functions. 
Is it possible to break the offending function into several simpler functions?

Comment (daviddrysdale) [reply gometalinter.json]: Not really, no.  

(I am not at all convinced that the gocyclo check is of any real use -- a simple count of `if` statements is very primitive, and so has false "positives" from entirely linear functions that have error checking.)

Comment (daviddrysdale) [reply gometalinter.json]: (To put it another way, if [the error handling mechanism suggested for Go 2](https://go.googlesource.com/proposal/+/master/design/go2draft-error-handling.md) were present, the cyclo count for the function in question would drop from 27 to ~1 with no change in the function's behaviour.)

File: docs/MapHashers.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+tree, but also allows for a more efficient Map API.
+
+In particular, any values in a leaf inclusion proof that match the empty hash
+value for that node need not be sent on the wire.  Instead, a `nil` value is used
+to indicate that this node in the proof is the top of an empty subtree (a
+subtree all of whose leaves are empty / never-been-set).
+
+Also, the distinction between empty / never-set leaves, and leaves whose values
+have been explicitly set to a zero-length value, means that two possible hash
+values apply for leaves with `len(leaf.LeafValue)==0`.
```
Comment (gdbelvin) [inline docs/MapHashers.md]: Which can be distinguished with the value of `leaf.LeafHash`.  `len(leaf.LeafHash)==0` means that the value is empty. `len(leaf.LeafHash)!=0` means that `leaf.LeafValue` has been committed to explicitly, even if  `len(leaf.LeafValue) == 0`.

File: integration/map_integration_test.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/google/trillian/testonly"
+	"github.com/google/trillian/testonly/integration"
+
+	_ "github.com/google/trillian/crypto/keys/der/proto" // Register PrivateKey ProtoHandler
+	stestonly "github.com/google/trillian/storage/testonly"
+)
+
+var singleTX = flag.Bool("single_transaction", false, "Experimental: whether to update the map in a single transaction")
+
+func TestInProcessMapIntegration(t *testing.T) {
```
Comment (gdbelvin) [inline integration/map_integration_test.go]: `integration/maptest/map_test.go` also defines an in-process map integration test harness. 
Do you think this test could be refactored in that style?  It would enable this test to be reused for remote map implementations as well.

File: merkle/hashers/tree_hasher.go [hunk 8]
```diff
@@ -39,9 +39,16 @@ type MapHasher interface {
 	// A height of 0 indicates an empty leaf. The maximum height is Size*8.
 	// TODO(gbelvin) fully define index.
 	HashEmpty(treeID int64, index []byte, height int) []byte
-	// HashLeaf computes the hash of a leaf that exists.
+	// HashLeaf computes the hash of a leaf that exists.  This method
+	// is *not* used for computing the hash of a leaf that does not exist
+	// (instead, HashEmpty(treeID, index, 0) is used), as the hash value
+	// can be different between:
+	//  - a leaf that was previously set but now has been set to a zero
```
Comment (gdbelvin) [inline merkle/hashers/tree_hasher.go]: can be different between
- A leaf that is unset. 
- A leaf explicitly set, including set to []byte{}.

File: merkle/map_verifier.go [hunk 9]
```diff
@@ -28,9 +29,9 @@ import (
 // The process is essentially the same as the inclusion proof checking for
 // append-only logs, but adds support for nil/"default" proof nodes.
 //
-// Returns nil on a successful verification, and an error otherwise.
-func VerifyMapInclusionProof(treeID int64, index, leaf, expectedRoot []byte, proof [][]byte, h hashers.MapHasher) error {
-	if got, want := len(index)*8, h.BitLen(); got != want {
+// ReturnVs nil on a successful verification, and an error otherwise.
```
Comment (gdbelvin) [inline merkle/map_verifier.go]: Returns

File: server/map_rpc_server.go [hunk 10]
```diff
@@ -148,10 +148,7 @@ func (t *TrillianMapServer) getLeavesByRevision(ctx context.Context, mapID int64
 			found++
 		} else {
 			// Empty leaf for proof of non-existence.
-			leafHash, err := hasher.HashLeaf(mapID, index, nil)
-			if err != nil {
-				return nil, fmt.Errorf("HashLeaf(nil): %v", err)
-			}
+			leafHash := hasher.HashEmpty(mapID, index, 0)
```
Comment (gdbelvin) [inline server/map_rpc_server.go]: I think we decided to return "" for this case, in which we can just drop this line.

We should probably leave a comment to that effect though

File: server/map_rpc_server.go [hunk 11]
```diff
@@ -247,12 +244,6 @@ func (t *TrillianMapServer) SetLeaves(ctx context.Context, req *trillian.SetMapL
 			if err := checkIndexSize(l.Index, hasher); err != nil {
 				return err
 			}
-			if l.LeafValue == nil {
-				// Leaves are empty by default. Do not allow clients to store
-				// empty leaf values as this messes up the calculation of empty
-				// branches.
-				continue
-			}
 			leafHash, err := hasher.HashLeaf(mapID, l.Index, l.LeafValue)
```
Comment (gdbelvin) [inline server/map_rpc_server.go]: Might be worth adding a test that `HashLeaf(nil) == HashLeaf([]byte{})`

File: testonly/mapcontents.go [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+type ErrInvariant struct {
+	msg string
+}
+
+func (e ErrInvariant) Error() string {
+	return fmt.Sprintf("Invariant check failed: %v", e.msg)
+}
+
+// MapContents is a complete copy of the map's contents at a particular revision.
+type MapContents struct {
```
Comment (gdbelvin) [inline testonly/mapcontents.go]: Is this an in-memory implementation of the MapServer? 
Not sure if it would be useful to have it match the MapServer api with protos and all

Comment (gdbelvin) [reply testonly/mapcontents.go]: We could send each request to both implementations and check `proto.Equals` on the results.

File: gometalinter.json (hunk 5 above)
Comment (daviddrysdale) [reply gometalinter.json]: Shifted to a `// nolint: gocyclo`.

File: merkle/hashers/tree_hasher.go (hunk 8 above)
Comment (daviddrysdale) [reply merkle/hashers/tree_hasher.go]: Done.

File: merkle/map_verifier.go (hunk 9 above)
Comment (daviddrysdale) [reply merkle/map_verifier.go]: Done.

File: testonly/mapcontents.go (hunk 12 above)
Comment (daviddrysdale) [reply testonly/mapcontents.go]: No, `MapContents` is just a wrapper around a Go `map` to hold a mirror of the expected map contents at a revision.   (It's also pre-existing code, just extracted from the maphammer for re-use here.)

File: server/map_rpc_server.go (hunk 10 above)
Comment (daviddrysdale) [reply server/map_rpc_server.go]: Yeah, it's gone with the update to plan A revision 2.

File: docs/MapHashers.md (hunk 6 above)
Comment (daviddrysdale) [reply docs/MapHashers.md]: Tweaked.

File: integration/map_integration_test.go (hunk 7 above)
Comment (daviddrysdale) [reply integration/map_integration_test.go]: Possibly in a different PR, but I don't think this one quite matches up with the table-driven tests in there.

**Comment by daviddrysdale** [comment]
Updated to use `leaf.LeafHash=nil` for empty leaves.

File: testonly/mapcontents.go [hunk 13]
```diff
@@ -127,6 +129,114 @@ func (m *MapContents) UpdatedWith(rev uint64, leaves []*trillian.MapLeaf) *MapCo
 	return &result
 }
 
+type bitString struct {
```
Comment (AlCutter) [inline testonly/mapcontents.go]: `big.Int` _might_ do a lot of what you want from this?

File: trillian_map_api.proto [hunk 14]
```diff
@@ -44,6 +51,10 @@ message MapLeaf {
 
 message MapLeafInclusion {
   MapLeaf leaf = 1;
+  // inclusion holds the inclusion proof for this leaf in the map root. It
+  // holds one entry for each level of the tree; combining each of these in
+  // turn with the leaf's hash (according to the tree's hash strategy)
+  // reproduces the root hash.
```
Comment (AlCutter) [inline trillian_map_api.proto]: Worth adding a bit about `nil`s in the inclusion proof in the "undefined" areas?

File: integration/map.go [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+		if err := mi.verifier.VerifyMapLeafInclusionHash(root.RootHash, leavesRsp.MapLeafInclusion[i]); err != nil {
+			return fmt.Errorf("failed to verify inclusion proof for key=%x: %v", key, err)
+		}
+	}
+	return nil
+}
+
+// expectValue retrieves the current map value for a particular key and checks that the
+// associated LeafValue is the specified value.
+func (mi *mapInfo) expectValue(ctx context.Context, key, value []byte) error {
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply integration/map.go]: Done.

File: integration/map.go [hunk 16]
```diff
@@ -0,0 +1,311 @@
...
+	fmt.Printf("%d:   Leaf[%x] = %q\n", mi.id, key, got)
+	if !bytes.Equal(got, value) {
+		return fmt.Errorf("unexpected leaf value %x from get-map-leaves", got)
+	}
+	return nil
+}
+
+// timeFromNanos converts a timestamp in nanoseconds  to a time.Time.
+func timeFromNanos(nanos uint64) time.Time {
+	secs := int64(nanos / 1000000000)
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply integration/map.go]: It ends up being `time.Unix(0, int64(nanos))` which is a smidgeon longer, so I've left `timeFromNanos` in place and just simplified its implementation.

File: integration/map_integration_test.go [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+	defer env.Close()
+
+	tree, err := client.CreateAndInitTree(ctx, &trillian.CreateTreeRequest{Tree: stestonly.MapTree}, env.Admin, env.Map, nil)
+	if err != nil {
+		t.Fatalf("Failed to create log: %v", err)
+	}
+
+	verifier, err := client.NewMapVerifierFromTree(tree)
+	if err != nil {
+		t.Fatalf("failed to create map verifier: %v", err)
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply integration/map_integration_test.go]: Done.

File: testonly/mapcontents.go [hunk 18]
```diff
@@ -127,6 +129,114 @@ func (m *MapContents) UpdatedWith(rev uint64, leaves []*trillian.MapLeaf) *MapCo
...
+	data   [sha256.Size]byte
+	bitLen int
+}
+
+var topBitMask = map[int]byte{0: 0x00, 1: 0x80, 2: 0xc0, 3: 0xe0, 4: 0xf0, 5: 0xf8, 6: 0xfc, 7: 0xfe}
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply testonly/mapcontents.go]: Done.

File: trillian_map_api.proto (hunk 14 above)
Comment (daviddrysdale) [reply trillian_map_api.proto]: Yeah, I meant to do that.

File: testonly/mapcontents.go (hunk 13 above)
Comment (daviddrysdale) [reply testonly/mapcontents.go]: It would do single-bit setting, but would make clearing suffixes harder and is also a bit misleading (this value is never interpreted as an integer) so I'd prefer to stick as-is.
